cpp/oneapi/dal/backend/primitives/reducer.cpp

```diff
@@ -0,0 +1,324 @@
+
+template <typename Float>
+struct binary_functor<Float, binary_operation::sum> {
+    constexpr static Float init_value = 0;
+
+    Float operator()(Float a, Float b) const {
+        return (a + b);
+    }
+};
+
+template <typename Float>
+struct binary_functor<Float, binary_operation::mul> {
+    constexpr static Float init_value = 1;
+
+    Float operator()(Float a, Float b) const {
+        return (a * b);
+    }
+};
+
+template <typename Float>
+struct binary_functor<Float, binary_operation::max> {
+    constexpr static Float init_value = std::numeric_limits<Float>::min();
+
+    Float operator()(Float a, Float b) const {
+        return std::max(a, b);
+    }
+};
+
+template <typename Float>
+struct binary_functor<Float, binary_operation::min> {
+    constexpr static Float init_value = std::numeric_limits<Float>::max();
+
+    Float operator()(Float a, Float b) const {
+        return std::min(a, b);
+    }
+};
+
+//Direct instatiation
+template struct binary_functor<float, binary_operation::sum>;
+template struct binary_functor<double, binary_operation::sum>;
+template struct binary_functor<float, binary_operation::mul>;
+template struct binary_functor<double, binary_operation::mul>;
+template struct binary_functor<float, binary_operation::min>;
+template struct binary_functor<double, binary_operation::min>;
+template struct binary_functor<float, binary_operation::max>;
+template struct binary_functor<double, binary_operation::max>;
+
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+
+namespace impl {
+
+template <unary_operation UnOp, binary_operation BinOp, typename Float, bool IsRowMajorLayout>
+struct reducer_singlepass_kernel {
+public:
+    //Zero cost
+    const binary_functor<Float, BinOp> binary;
+    const unary_functor<Float, UnOp> unary;
+
+protected:
+    typedef cl::sycl::
+        accessor<Float, 1, cl::sycl::access::mode::read_write, cl::sycl::access::target::local>
+            local_acc_t;
+
+public:
+    reducer_singlepass_kernel(std::int64_t vector_size_,
+                              size_t n_vectors_,
+                              const Float* vectors_,
+                              Float* reduces_,
+                              local_acc_t partial_reduces_)
+            : vector_size(vector_size_),
+              n_vectors(n_vectors_),
+              vectors(vectors_),
+              reduces(reduces_),
+              partial_reduces(partial_reduces_){};
+
+    void operator()(cl::sycl::nd_item<2> idx) {
+        const std::uint32_t local_size = idx.get_local_range(0);
+
+        const std::uint32_t global_dim = IsRowMajorLayout ? vector_size : 1;
+        const std::uint32_t local_dim  = IsRowMajorLayout ? 1 : n_vectors;
+
+        const std::uint32_t item_id  = idx.get_local_id(0);
+        const std::uint32_t group_id = idx.get_global_id(1);
+
+        Float el                 = vectors[group_id * global_dim + item_id * local_dim];
+        partial_reduces[item_id] = binary.init_value;
```

Comment:
You could consider another option: reduction inside subgroups could be done by subgroup operations; each subgroup stores its value in the array (array size = number of subgroups); at the end single (first) subgroup reads the array and performs final reduction.

Reply:
It will be explored in the future. Thank you for this significant remark